File: typescript/argumentholder.ts
import * as crypto from 'crypto'
import * as path from 'path'

export class ArgumentHolder{
    directoryA: string = "";
    directoryB: string = "";

    checksumName: string = "";
    checksumGenerator: Function = () => {};

    ignoreUnchanged: boolean = false;

    public directoryInfoIsMissing(){
        return typeof this.directoryA == "undefined" || typeof this.directoryB == "undefined"
            || this.directoryA.length < 3 || this.directoryB.length < 3
    }

    public verifyArguments(args: any){
        this.ignoreUnchanged = args["ignoreUnchanged"] || false
        this.selectChecksum(args);
        this.directoryA = this.regularizeDirectoryName(this.directoryA)
        this.directoryB = this.regularizeDirectoryName(this.directoryB)
    }

    private regularizeDirectoryName(somePath: string){
        let cleanPath = path.normalize(somePath)
        if(cleanPath.lastIndexOf('/') == cleanPath.length-1){
            cleanPath = cleanPath.substring(0, cleanPath.length-1)
        }

        return cleanPath
    }

    private selectChecksum(args: any) {
        const options = ['md5', 'sha1', 'sha256'];
        const notImplemented = ['crc32', 'adler32'];
        let selection = "";

        notImplemented.forEach(option =>{
            if(args[option]){
                throw new Error(`Cannot accept checksum '${option}'. It has not been implemented yet`);
            }
        });

        options.forEach(functionName => {
            let possibleSelection = args[functionName];
            if (possibleSelection) {
                if(selection.length > 1){
                    throw new Error(`Expected only option '${selection}' or '${functionName}'`);
                }

                selection = functionName;
            }
        });
        
        if (selection.length < 1) {
            selection = 'md5';
        }
        this.checksumName = selection;
    }

    public getChecksumObject(){
        return crypto.createHash(this.checksumName)
    }
}

File: golang/argument_holder.go
package main

import (
	"crypto/md5"
	"crypto/sha1"
	"crypto/sha256"
	"hash"
	"hash/adler32"
	"hash/crc32"
	"path/filepath"

	kingpin "gopkg.in/alecthomas/kingpin.v2"
)

// ArgumentHolder holds the arguments as parsed and determines the checksum name
type ArgumentHolder struct {
	directoryA      *string
	directoryB      *string
	ignoreUnchanged *bool

	// Set by the function below
	checksumName string
}

func (args *ArgumentHolder) validate(app *kingpin.Application, optionMap map[string]*bool) {
	chosenChecksum := ""
	for hashName, optionSelected := range optionMap {
		if !(*optionSelected) {
			continue
		}

		if chosenChecksum != "" {
			app.FatalUsage("Expected only option '%s' or '%s'\n", chosenChecksum, hashName)
		}

		chosenChecksum = hashName
	}

	if chosenChecksum == "" {
		chosenChecksum = "md5"
	}

	args.checksumName = chosenChecksum

	// Also do directory clean up
	*args.directoryA = filepath.Clean(*args.directoryA)
	*args.directoryB = filepath.Clean(*args.directoryB)
}

type checksumGenerator func() hash.Hash

func (args *ArgumentHolder) getChecksumInstance() hash.Hash {
	return checksums[args.checksumName]()
}

func getAdler32() hash.Hash {
	return adler32.New().(hash.Hash)
}

func getCRC32() hash.Hash {
	return crc32.NewIEEE().(hash.Hash)
}

var checksums = map[string]checksumGenerator{
	"md5":     md5.New,
	"sha1":    sha1.New,
	"sha256":  sha256.New,
	"adler32": getAdler32,
	"crc32":   getCRC32,
}
